### ISSUE TITLE
Error on recalc today - 'nothing to repeat'

### ISSUE THREAD

**Issue by ashprice** [mortii/anki-morphs#159]
**Describe the bug**
I get an error at the end of the recalc process, after the database has been saved and the cards updated. Anki then shows incorrect morph amounts in the UI, though if I restart Anki, it reads them correctly from the db. 

Not sure if card ordering is updated correctly or not, will look later today. I'll also poke around in parser.py to see if I can spot the issue myself. 

**To Reproduce**
1. Press recalc.
2. Don't profit. 

My configuration should be fine as it was working before I think some commits pushed today or yesterday. 

Here is the error, with my OS and Anki version info:

```
Anki 23.12.1 (68b374e6) (src) (ao)
Python 3.11.7 Qt 6.6.2 PyQt 6.6.1
Platform: Linux-6.7.5-arch1-1-x86_64-with-glibc2.39

Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/aqt/taskman.py", line 142, in _on_closures_pending
    closure()
  File "/usr/lib/python3.11/site-packages/aqt/taskman.py", line 86, in <lambda>
    lambda future: self.run_on_main(lambda: on_done(future))
                                            ^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/aqt/taskman.py", line 106, in wrapped_done
    on_done(fut)
  File "/usr/lib/python3.11/site-packages/aqt/operations/__init__.py", line 252, in wrapped_done
    self._failure(exception)
  File "/home/hearth/.local/share/Anki2/addons21/472573498/recalc.py", line 914, in _on_failure
    raise error
  File "/usr/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/aqt/operations/__init__.py", line 242, in wrapped_op
    return self._op(mw.col)
           ^^^^^^^^^^^^^^^^
  File "/home/hearth/.local/share/Anki2/addons21/472573498/recalc.py", line 77, in _recalc_background_op
    _update_cards_and_notes(am_config)
  File "/home/hearth/.local/share/Anki2/addons21/472573498/recalc.py", line 445, in _update_cards_and_notes
    _update_highlighted_field(
  File "/home/hearth/.local/share/Anki2/addons21/472573498/recalc.py", line 788, in _update_highlighted_field
    highlighted_text = text_highlighting.get_highlighted_text(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hearth/.local/share/Anki2/addons21/472573498/text_highlighting.py", line 59, in get_highlighted_text
    span_elements, text_to_highlight = _extract_span_elements_and_filter_string(
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hearth/.local/share/Anki2/addons21/472573498/text_highlighting.py", line 189, in _extract_span_elements_and_filter_string
    morph_matches = re.finditer(regex_pattern, text_to_highlight)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/__init__.py", line 223, in finditer
    return _compile(pattern, flags).finditer(string)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/__init__.py", line 294, in _compile
    p = _compiler.compile(pattern, flags)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/_compiler.py", line 745, in compile
    p = _parser.parse(p, flags)
        ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/_parser.py", line 989, in parse
    p = _parse_sub(source, state, flags & SRE_FLAG_VERBOSE, 0)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/_parser.py", line 464, in _parse_sub
    itemsappend(_parse(source, state, verbose, nested + 1,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/re/_parser.py", line 691, in _parse
    raise source.error("nothing to repeat",
re.error: nothing to repeat at position 0
```

If I had to guess, the relevant commit is probably the ruby one? Maybe some brackets or HTML or something in a card is the issue. Not sure how I'll track it down in a library my size...

**Comment by mortii** [comment]
Oh, that sucks. 

How large is your collection? I don't have access to a computer at the moment, but iirc there is a file called 'collection.anki2'  in the profile folder that has the necessery data (no media files), so if you upload that file and your ankimorphs settings (tools -> addons -> AnkiMorphs -> config, copy paste the info), then I could try to debug it.

Also, i forgot to add to the guide that the preprocess option 'ignore everything in brackets []' has to be selected for the furigana option to work, so there might be something there that is causing problems.

**Comment by ashprice** [comment]
Hmm, my collection has become quite large as I have some mining decks in there, around 122MB. I'd feel a bit sheepish about uploading the whole thing as I have some more personal stuff in my DB (I use a personal sync server), so I went through to try to isolate the issue. 

As of now, I have at least one hit, which is this deck of Korean cards I downloaded (I am not actually learning Korean _really_, I got it mostly for the sake of learning a few words/phrases.) 

Here is a colpkg without media: https://we.tl/t-rdP3YhQaqq

Sorry for using wetransfer, github wouldn't accept the file size. Also, not sure if the older format is preferable. 

Looking at it myself, I'm assuming the error is because of the silly whitespace and [ ] and stuff a lot of these cards have. For example this one card has a back field like this hahah:

```
백











[Sino-Korean #]
```

I'll eventually edit all of this out but yeah...

I think the only thing that is relevant in my config is that I pointed Ankimorphs at the 'Back' field, and do indeed have the preprocessing option for square brackets enabled. 

I assume it isn't relevant, but here is my configuration for Ankimorphs generally + this card type (I stripped other card types out):

```
{
    "extra_difficulty": true,
    "extra_highlighted": true,
    "extra_unknowns": true,
    "extra_unknowns_count": true,
    "filters": [
    ... other card types here ...
        {
            "extra_difficulty": true,
            "extra_highlighted": true,
            "extra_unknowns": true,
            "extra_unknowns_count": true,
            "field": "Back",
            "field_index": 1,
            "modify": true,
            "morph_priority": "Collection frequency",
            "morph_priority_index": 0,
            "morphemizer_description": "spaCy: ko_core_news_lg",
            "morphemizer_name": "SpacyMorphemizer",
            "note_type": "Korean Detailed Vocabulary (with reversed card)",
            "note_type_id": 1627228057054,
            "read": true,
            "tags": {
                "exclude": [],
                "include": []
            }
        }
    ],
    "preprocess_ignore_bracket_contents": true,
    "preprocess_ignore_names_morphemizer": false,
    "preprocess_ignore_names_textfile": false,
    "preprocess_ignore_round_bracket_contents": false,
    "preprocess_ignore_slim_round_bracket_contents": false,
    "preprocess_ignore_suspended_cards_content": false,
    "recalc_interval_for_known": 21,
    "recalc_on_sync": false,
    "recalc_read_known_morphs_folder": false,
    "recalc_suspend_known_new_cards": false,
    "recalc_toolbar_stats_use_known": true,
    "recalc_toolbar_stats_use_seen": false,
    "recalc_unknowns_field_shows_inflections": true,
    "recalc_unknowns_field_shows_lemmas": false,
    "shortcut_browse_all_same_unknown": "Shift+L",
    "shortcut_browse_ready_same_unknown": "L",
    "shortcut_browse_ready_same_unknown_lemma": "Ctrl+Shift+L",
    "shortcut_frequency_file_generator": "Ctrl+Shift+F",
    "shortcut_known_morphs_exporter": "Ctrl+Shift+E",
    "shortcut_learn_now": "Ctrl+Alt+N",
    "shortcut_readability_report_generator": "Ctrl+Shift+R",
    "shortcut_recalc": "Ctrl+M",
    "shortcut_set_known_and_skip": "K",
    "shortcut_settings": "Ctrl+O",
    "shortcut_view_morphemes": "Ctrl+Alt+V",
    "skip_only_known_morphs_cards": false,
    "skip_show_num_of_skipped_cards": true,
    "skip_unknown_morph_seen_today_cards": false,
    "tag_known_automatically": "am-known-automatically",
    "tag_known_manually": "am-known-manually",
    "tag_learn_card_now": "am-learn-card-now",
    "tag_not_ready": "am-not-ready",
    "tag_ready": "am-ready"
}
```

I will continue going through my card types to see if I've any other hits (almost done).

**Comment by mortii** [comment]
Nice, i'll take a look when I get back to a computer. Based off of your example im guessing the regex fails on some special characters or something. When reworking the highlighting I forgot to catch regex exceptions (the previous version did), so this might be a quick fix.

**Comment by xofm31** [comment]
I think I had the same problem, with cards that had parentheses. Here is an example of the text on the card: "智 (慧)"

**Comment by mortii** [comment]
@xofm31 thanks for letting me know! 

I'm back at a computer so I'll take a look at this now.

v1.2.2 todo list:
- [ ] fix this bug
- [ ] guide
    - [ ] specify that 'ignore brackets' option has to be activated for ruby characters to work
    - [ ] update the images in known problems to include different highlighting of the effected morphs in the furigana problem

**Comment by mortii** [comment]
This was an easy fix. Sometimes morphs include special regex characters for whatever reason, so just wrapping the regex search pattern in re.escape() prevents this crash.

@ashprice for your note filter it crashed on this morph: `?몇`, from the text: 
```html
몇...?<div><br></div><div>몇...</div>
```
These malformed morph occur when \<br\> is eaten by the parser. The 'morphs don't split on punctuation marks" section in [known problem](https://mortii.github.io/anki-morphs/user_guide/known-problems.html) goes into details on how to prevent this.

@xofm31 I'm not able to reproduce the problem with your example, but maybe it will go away with my upcoming hotfix, if not, let me know!

**Comment by mortii** [comment]
Released [v1.2.2](https://github.com/mortii/anki-morphs/releases/tag/v1.2.2)! 

Let know if it still persists or if you encounter other problems.

**Comment by mortii** [comment]
@xofm31 Oh, very interesting. I'll create a new issue for it.
